sled-agent/src/storage_manager.rs

```diff
@@ -574,6 +581,12 @@ impl StorageWorker {
         resources: &StorageResources,
         disk: UnparsedDisk,
     ) -> Result<(), Error> {
+        // We need to encrypt external disks, and so we wait until
+        // the storage key requester is ready.
+        if disk.variant() == DiskVariant::U2 && !self.key_requester.is_ready() {
+            return Ok(());
+        }
```

Comment:
Question for you here, and anywhere else we check `!self.key_request.is_ready()` before maybe returning `Ok(())` -- what causes us to "scan again" once the key requester is ready?

Here's what I vaguely fear:

- Notification to upsert disk A
- Key request is marked as ready
- We never use disk A, because we **think** we upserted it, but in reality we just ignored it

Reply:
Great question! I was also concerned about this, but I think it's fine from my reading of the code. Please correct me if I'm wrong.

We set the readiness at the [beginning of `request_sled_agent`](https://github.com/oxidecomputer/omicron/blob/main/bootstore/src/schemes/v0/storage.rs#L58-L87). `request_sled_agent` always gets called in the `Before` state on cold boot.  The other important time `request_sled_agent` gets called is via RSS init. During that call we also start in before. The only time I think we start in `After` is when RSS reset gets called.

In the `Before` state we always call [`SledServer::start`](https://github.com/oxidecomputer/omicron/blob/main/bootstore/src/schemes/v0/storage.rs#L58-L87) which calls `SledAgent::new` which ends up [starting a new hardware monitor task](https://github.com/oxidecomputer/omicron/blob/09c6faab7770be71d62d5dd91ccd75a36d87b6a5/sled-agent/src/sled_agent.rs#L365-L367) which does a [`full_hardware_scan`](https://github.com/oxidecomputer/omicron/blob/09c6faab7770be71d62d5dd91ccd75a36d87b6a5/sled-agent/src/sled_agent.rs#L421) which I believe will upsert all the disks.

However, you now have me concerned about the opposite problem. When we do an RSS reset I don't think we unset the readiness. I'll need to look into this.

Reply:
I still think this is pretty dependent on "the code is right because the thing a module away from me is implicitly happening to call me in a peculiar way". Silently opting to ignore a disk feels like it's bug-prone behavior, especially because there's no indication to the caller that we **should** re-scan.

As one example of this: I don't think your example works with synthetic disks, which get injected once?

Backing up a sec, what's the **point** of this "key_requester.is_ready" stuff? It's supposed to basically tell us to "back off" on loading U.2s until we've gotten far enough that we can successfully put the `crypt` dataset in place, right?

Take a look at `do_work_internal` in this file -- maybe it makes sense to simply track all added U.2s in some sort of "deferred queue" that gets unloaded once this variable flips to true? (Basically, just push these disks onto a HashSet before returning in all these spots). Then the caller wouldn't need to know to re-scan, **and** we'd handle the synthetic disk case.

Reply:
Ooh. Great idea! I like this much better. This was just the first way to do it that I thought of yesterday, and I agree with you that it is full of peril. Very much +1 to a more straightforward and obviously correct path!

Reply:
After taking a closer look at your proposal I think we'll have to add another `StorageWorkerRequest` to inform it to go ahead and actually upsert the disks it had queued after the system is ready. We also may want to pass the `HashSet` in mutably from `do_work` so that if there is an error and retry we don't lose the disks.

Reply:
Implemented in [2151d7f](https://github.com/oxidecomputer/omicron/pull/3652/commits/2151d7fba79c5729286c138dbb35edfdc28824fe)